server/server.go

```diff
@@ -1147,3 +1156,188 @@ func (s *Server) getClientConnectURLs() []string {
 
 	return urls
 }
+
+// if the ip is not specified, attempt to resolve it
+func resolveHostPorts(addr net.Listener) []string {
+	hostPorts := make([]string, 0)
+	hp := addr.Addr().(*net.TCPAddr)
+	port := strconv.Itoa(hp.Port)
+	if hp.IP.IsUnspecified() {
+		var ip net.IP
+		ifaces, _ := net.Interfaces()
+		for _, i := range ifaces {
+			addrs, _ := i.Addrs()
+			for _, addr := range addrs {
+				switch v := addr.(type) {
+				case *net.IPNet:
+					ip = v.IP
+					hostPorts = append(hostPorts, net.JoinHostPort(ip.String(), port))
+				case *net.IPAddr:
+					ip = v.IP
+					hostPorts = append(hostPorts, net.JoinHostPort(ip.String(), port))
+				default:
+					continue
+				}
+			}
+		}
+	} else {
+		hostPorts = append(hostPorts, net.JoinHostPort(hp.IP.String(), port))
+	}
+	return hostPorts
+}
+
+// format the address of a net.Listener with a protocol
+func formatURL(protocol string, addr net.Listener) []string {
+	hostports := resolveHostPorts(addr)
+	for i, hp := range hostports {
+		hostports[i] = fmt.Sprintf("%s://%s", protocol, hp)
+	}
+	return hostports
+}
+
+// Ports describes URLs that the server can be contacted in
+type Ports struct {
+	Nats       []string `json:"nats,omitempty"`
+	Monitoring []string `json:"monitoring,omitempty"`
+	Cluster    []string `json:"cluster,omitempty"`
+	Profile    []string `json:"profile,omitempty"`
+}
+
+// Returns a Ports struct describing ports where the server can be contacted
+func (s *Server) PortsInfo(opts *Options) Ports {
+	s.mu.Lock()
+	info := s.copyInfo()
+	listener := s.listener
+	httpListener := s.http
+	clusterListener := s.routeListener
+	profileListener := s.profiler
+	s.mu.Unlock()
+
+	ports := Ports{}
+
+	if listener != nil {
+		natsProto := "nats"
+		if info.TLSRequired {
+			natsProto = "tls"
+		}
+		ports.Nats = formatURL(natsProto, listener)
+	}
+
+	if httpListener != nil {
+		monProto := "http"
+		if opts.HTTPSPort != 0 {
+			monProto = "https"
+		}
+		ports.Monitoring = formatURL(monProto, httpListener)
+	}
+
+	if clusterListener != nil {
+		clusterProto := "nats"
+		if opts.Cluster.TLSConfig != nil {
+			clusterProto = "tls"
+		}
+		ports.Cluster = formatURL(clusterProto, clusterListener)
+	}
+
+	if profileListener != nil {
+		ports.Profile = formatURL("http", profileListener)
+	}
+
+	return ports
+}
+
+// Returns the portsFile. If a non-empty dirHint is provided, the dirHint
+// path is used instead of the server option value
+func (s *Server) PortsFile(dirHint string) string {
+	dirname := s.getOpts().PortsFileDir
+	if dirHint != "" {
+		dirname = dirHint
+	}
+	if dirname == _EMPTY_ {
+		return _EMPTY_
+	}
+	return path.Join(dirname, fmt.Sprintf("%s_%d.ports", path.Base(os.Args[0]), os.Getpid()))
+}
+
+// Delete the ports file. If a non-empty dirHint is provided, the dirHint
+// path is used instead of the server option value
+func (s *Server) deletePortsFile(hintDir string) {
+	portsFile := s.PortsFile(hintDir)
```

Comment:
delete ports file should work with provided dir. This is used in the context of either reload where we pass the known old directory, or at shutdown where we will pass current server option's port file dir.

Reply:
Same as above